src/jgikbase/idmapping/storage/mongo/id_mapping_mongo_storage.py

```diff
@@ -231,7 +235,30 @@ def get_namespace(self, namespace_id: NamespaceID) -> Namespace:
             nsdoc[_FLD_PUB_MAP],
             self._to_user_set(nsdoc[_FLD_USERS]))
 
-    def _to_user_set(self, userdocs):
-        # TODO implement when add / remove user are implemented
-        userdocs.clear()
-        return set()
+    def _to_user_set(self, userdocs) -> Set[User]:
+        return {User(AuthsourceID(u[_FLD_AUTHSOURCE]), u[_FLD_NAME]) for u in userdocs}
+
+    def add_user_to_namespace(self, namespace_id: NamespaceID, admin_user: User) -> None:
+        self._modify_namespace_users(True, namespace_id, admin_user)
+
+    def remove_user_from_namespace(self, namespace_id: NamespaceID, admin_user: User) -> None:
+        self._modify_namespace_users(False, namespace_id, admin_user)
+
+    def _modify_namespace_users(self, add, namespace_id, admin_user):
+        not_none(namespace_id, 'namespace_id')
+        not_none(admin_user, 'admin_user')
+        op = '$addToSet' if add else '$pull'
+        try:
+            res = self._db[_COL_NAMESPACES].update_one(
+                {_FLD_NS_ID: namespace_id.id},
+                {op: {_FLD_USERS: {_FLD_AUTHSOURCE: admin_user.authsource_id.id,
```

Comment:
Do you have to duplicate all the user fields in the namespace collection when it already exists in the users collection?

Reply:
None of them are actually the same string.